File: apps/desktop/src-tauri/src/core/router/providers/anthropic.rs
use crate::core::router::sse_parser::{parse_sse_stream, StreamChunk};
use crate::core::router::{
    ContentPart, ImageFormat, LLMProvider, LLMRequest, LLMResponse, ToolCall,
};
use futures_util::Stream;
use reqwest::Client;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::error::Error;
use std::pin::Pin;

#[derive(Debug, Clone, Serialize)]
struct AnthropicMessage {
    role: String,
    content: AnthropicMessageContent,
}

#[derive(Debug, Clone, Serialize)]
#[serde(untagged)]
enum AnthropicMessageContent {
    Text(String),
    Multimodal(Vec<AnthropicContentPart>),
}

#[derive(Debug, Clone, Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
enum AnthropicContentPart {
    Text { text: String },
    Image { source: AnthropicImageSource },
}

#[derive(Debug, Clone, Serialize)]
struct AnthropicImageSource {
    #[serde(rename = "type")]
    source_type: String,
    media_type: String,
    data: String,
}

#[derive(Debug, Clone, Serialize)]
struct AnthropicTool {
    name: String,
    description: String,
    input_schema: Value,
}

#[derive(Debug, Clone, Serialize)]
struct AnthropicThinking {
    #[serde(rename = "type")]
    thinking_type: String,
    budget_tokens: u32,
}

#[derive(Debug, Clone, Serialize)]
struct AnthropicRequest {
    model: String,
    messages: Vec<AnthropicMessage>,
    #[serde(skip_serializing_if = "Option::is_none")]
    system: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    max_tokens: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    temperature: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    stream: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    tools: Option<Vec<AnthropicTool>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    thinking: Option<AnthropicThinking>,
}

#[derive(Debug, Clone, Deserialize)]
struct AnthropicResponse {
    _id: String,
    content: Vec<AnthropicContent>,
    usage: AnthropicUsage,
    model: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    stop_reason: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
enum AnthropicContent {
    Text {
        text: String,
    },
    ToolUse {
        id: String,
        name: String,
        input: Value,
    },
}

#[derive(Debug, Clone, Deserialize)]
struct AnthropicUsage {
    input_tokens: u32,
    output_tokens: u32,
}

pub struct AnthropicProvider {
    api_key: String,
    client: Client,
    base_url: String,
}

impl AnthropicProvider {
    pub fn new(api_key: String) -> Self {
        Self {
            api_key,
            client: Client::new(),
            base_url: "https://api.agiworkforce.com".to_string(),
        }
    }

    fn convert_content(
        text: &str,
        multimodal: Option<&Vec<ContentPart>>,
    ) -> AnthropicMessageContent {
        if let Some(parts) = multimodal {
            let mut anthropic_parts = Vec::new();

            if !text.is_empty() {
                anthropic_parts.push(AnthropicContentPart::Text {
                    text: text.to_string(),
                });
            }

            for part in parts {
                match part {
                    ContentPart::Text { text } => {
                        anthropic_parts.push(AnthropicContentPart::Text { text: text.clone() });
                    }
                    ContentPart::Image { image } => {
                        let media_type = match image.format {
                            ImageFormat::Png => "image/png",
                            ImageFormat::Jpeg => "image/jpeg",
                            ImageFormat::Webp => "image/webp",
                        };
                        let base64_data = base64::Engine::encode(
                            &base64::engine::general_purpose::STANDARD,
                            &image.data,
                        );
                        anthropic_parts.push(AnthropicContentPart::Image {
                            source: AnthropicImageSource {
                                source_type: "base64".to_string(),
                                media_type: media_type.to_string(),
                                data: base64_data,
                            },
                        });
                    }
                    ContentPart::Video { .. } => {}
                }
            }

            AnthropicMessageContent::Multimodal(anthropic_parts)
        } else {
            AnthropicMessageContent::Text(text.to_string())
        }
    }

    fn calculate_cost(model: &str, input_tokens: u32, output_tokens: u32) -> f64 {
        let (input_cost, output_cost) = match model {
            "claude-sonnet-4-5" | "claude-4.5-sonnet" => (3.0, 15.0),
            "claude-haiku-4-5" | "claude-4.5-haiku" => (0.5, 2.5),
            "claude-opus-4-5" | "claude-4.5-opus" => (12.0, 36.0),

            "claude-opus-4" | "claude-4-opus" => (15.0, 75.0),

            "claude-3-5-sonnet-20241022" | "claude-3-5-sonnet-latest" => (3.0, 15.0),
            "claude-3-5-haiku-20241022" | "claude-3-5-haiku-latest" => (0.80, 4.0),

            "claude-3-opus-20240229" => (15.0, 75.0),
            "claude-3-sonnet-20240229" => (3.0, 15.0),
            "claude-3-haiku-20240307" => (0.25, 1.25),

            _ => (3.0, 15.0),
        };

        let input = (input_tokens as f64 / 1_000_000.0) * input_cost;
        let output = (output_tokens as f64 / 1_000_000.0) * output_cost;
        input + output
    }
}

#[async_trait::async_trait]
impl LLMProvider for AnthropicProvider {
    async fn send_message(
        &self,
        request: &LLMRequest,
    ) -> Result<LLMResponse, Box<dyn Error + Send + Sync>> {
        let anthropic_tools = request.tools.as_ref().map(|tools| {
            tools
                .iter()
                .map(|tool| AnthropicTool {
                    name: tool.name.clone(),
                    description: tool.description.clone(),
                    input_schema: tool.parameters.clone(),
                })
                .collect()
        });

        let system_message = request
            .messages
            .iter()
            .filter(|m| m.role == "system")
            .map(|m| m.content.clone())
            .collect::<Vec<String>>()
            .join("\n\n");

        let system_param = if system_message.is_empty() {
            None
        } else {
            Some(system_message)
        };

        let messages: Vec<AnthropicMessage> = request
            .messages
            .iter()
            .filter(|m| m.role != "system")
            .map(|m| AnthropicMessage {
                role: m.role.clone(),
                content: Self::convert_content(&m.content, m.multimodal_content.as_ref()),
            })
            .collect();

        let (thinking, temperature, max_tokens) = if request.thinking_mode.unwrap_or(false) {
            (
                Some(AnthropicThinking {
                    thinking_type: "enabled".to_string(),
                    budget_tokens: 16000,
                }),
                None,
                request.max_tokens.or(Some(64000)),
            )
        } else {
            (None, request.temperature, request.max_tokens.or(Some(4096)))
        };

        let anthropic_request = AnthropicRequest {
            model: request.model.clone(),
            messages,
            system: system_param,
            max_tokens,
            temperature,
            stream: if request.stream { Some(false) } else { None },
            tools: anthropic_tools,
            thinking,
        };

        let response = self
            .client
            .post(format!("{}/messages", self.base_url))
            .header("x-api-key", &self.api_key)
            .header("anthropic-version", "2023-06-01")
            .header("Content-Type", "application/json")
            .json(&anthropic_request)
            .send()
            .await?;

        if !response.status().is_success() {
            let status = response.status();
            let error_text = response
                .text()
                .await
                .unwrap_or_else(|_| "Unknown error".to_string());
            return Err(format!("Anthropic API error {}: {}", status, error_text).into());
        }

        let response_text = response.text().await?;
        tracing::debug!("Anthropic response body: {}", response_text);

        let anthropic_response: AnthropicResponse =
            serde_json::from_str(&response_text).map_err(|e| {
                format!(
                    "Failed to parse Anthropic response: {}. Body: {}",
                    e, response_text
                )
            })?;

        let mut text_content = String::new();
        let mut tool_calls = Vec::new();

        for content_block in &anthropic_response.content {
            match content_block {
                AnthropicContent::Text { text } => {
                    text_content.push_str(text);
                }
                AnthropicContent::ToolUse { id, name, input } => {
                    tool_calls.push(ToolCall {
                        id: id.clone(),
                        name: name.clone(),
                        arguments: serde_json::to_string(input).unwrap_or_default(),
                    });
                }
            }
        }

        let cost = Self::calculate_cost(
            &anthropic_response.model,
            anthropic_response.usage.input_tokens,
            anthropic_response.usage.output_tokens,
        );

        let total_tokens =
            anthropic_response.usage.input_tokens + anthropic_response.usage.output_tokens;

        let finish_reason =
            anthropic_response
                .stop_reason
                .as_ref()
                .map(|reason| match reason.as_str() {
                    "tool_use" => "tool_calls".to_string(),
                    "end_turn" => "stop".to_string(),
                    "max_tokens" => "length".to_string(),
                    _ => reason.clone(),
                });

        Ok(LLMResponse {
            content: text_content,
            tokens: Some(total_tokens),
            prompt_tokens: Some(anthropic_response.usage.input_tokens),
            completion_tokens: Some(anthropic_response.usage.output_tokens),
            cost: Some(cost),
            model: anthropic_response.model,
            tool_calls: if tool_calls.is_empty() {
                None
            } else {
                Some(tool_calls)
            },
            finish_reason,
            ..LLMResponse::default()
        })
    }

    fn is_configured(&self) -> bool {
        !self.api_key.is_empty() && self.api_key != "your-api-key-here"
    }

    fn name(&self) -> &str {
        "Anthropic"
    }

    fn supports_vision(&self) -> bool {
        true
    }

    fn supports_function_calling(&self) -> bool {
        true
    }

    async fn send_message_streaming(
        &self,
        request: &LLMRequest,
    ) -> Result<
        Pin<Box<dyn Stream<Item = Result<StreamChunk, Box<dyn Error + Send + Sync>>> + Send>>,
        Box<dyn Error + Send + Sync>,
    > {
        let anthropic_tools = request.tools.as_ref().map(|tools| {
            tools
                .iter()
                .map(|tool| AnthropicTool {
                    name: tool.name.clone(),
                    description: tool.description.clone(),
                    input_schema: tool.parameters.clone(),
                })
                .collect()
        });

        let system_message = request
            .messages
            .iter()
            .filter(|m| m.role == "system")
            .map(|m| m.content.clone())
            .collect::<Vec<String>>()
            .join("\n\n");

        let system_param = if system_message.is_empty() {
            None
        } else {
            Some(system_message)
        };

        let messages: Vec<AnthropicMessage> = request
            .messages
            .iter()
            .filter(|m| m.role != "system")
            .map(|m| AnthropicMessage {
                role: m.role.clone(),
                content: Self::convert_content(&m.content, m.multimodal_content.as_ref()),
            })
            .collect();

        let (thinking, temperature, max_tokens) = if request.thinking_mode.unwrap_or(false) {
            (
                Some(AnthropicThinking {
                    thinking_type: "enabled".to_string(),
                    budget_tokens: 16000,
                }),
                None,
                request.max_tokens.or(Some(64000)),
            )
        } else {
            (None, request.temperature, request.max_tokens.or(Some(4096)))
        };

        let anthropic_request = AnthropicRequest {
            model: request.model.clone(),
            messages,
            system: system_param,
            max_tokens,
            temperature,
            stream: Some(true),
            tools: anthropic_tools,
            thinking,
        };

        tracing::debug!(
            "Starting Anthropic streaming request for model: {}",
            request.model
        );

        let response = self
            .client
            .post(format!("{}/messages", self.base_url))
            .header("x-api-key", &self.api_key)
            .header("anthropic-version", "2023-06-01")
            .header("Content-Type", "application/json")
            .json(&anthropic_request)
            .send()
            .await?;

        if !response.status().is_success() {
            let status = response.status();
            let error_text = response
                .text()
                .await
                .unwrap_or_else(|_| "Unknown error".to_string());
            return Err(format!("Anthropic API error {}: {}", status, error_text).into());
        }

        tracing::debug!("Anthropic streaming response received, starting SSE parsing");

        Ok(Box::pin(parse_sse_stream(
            response,
            crate::core::router::Provider::Anthropic,
        )))
    }
}

File: apps/web/lib/llm-providers/anthropic.ts
import 'server-only';

import {
  BaseLLMProvider,
  LLMProviderRequest,
  LLMProviderResponse,
  RETRYABLE_HTTP_STATUS_CODES,
} from './base';
import { logger } from '@/lib/logger';

/**
 * Transform tools from OpenAI format to Anthropic format.
 * Handles four cases:
 *   1. Anthropic server-managed tools (web_search, code_execution) - pass through as-is
 *   2. Already-Anthropic custom tools (has input_schema) - pass through as-is
 *   3. OpenAI function format (type: 'function', function: {...}) - transform
 *   4. Bare format - transform with fallback input_schema
 */

/** Anthropic server-managed tool type prefixes (e.g. web_search_20260209, code_execution_20260120) */
const ANTHROPIC_SERVER_TOOL_PREFIXES = ['web_search_', 'code_execution_'];

interface OpenAITool {
  type?: string;
  function?: { name?: string; description?: string; parameters?: unknown };
  name?: string;
  description?: string;
  parameters?: unknown;
  input_schema?: unknown;
  // Server-managed tool fields (Anthropic-specific)
  [key: string]: unknown;
}

interface AnthropicTool {
  name?: string;
  type?: string;
  description?: string;
  input_schema?: unknown;
  [key: string]: unknown;
}

function transformTools(tools: OpenAITool[]): AnthropicTool[] {
  return tools.map((tool) => {
    // 1. Anthropic server-managed tools: type starts with a known server-tool prefix.
    //    These must be passed through exactly as-is - they have no input_schema.
    //    Examples: { type: 'web_search_20260209', name: 'web_search' }
    //             { type: 'code_execution_20260120', name: 'code_execution' }
    if (
      tool.type &&
      ANTHROPIC_SERVER_TOOL_PREFIXES.some((prefix) => (tool.type as string).startsWith(prefix))
    ) {
      return tool as unknown as AnthropicTool;
    }

    // 2. If tool is already in Anthropic custom tool format (has input_schema), use as-is
    if (tool.input_schema) {
      return tool as unknown as AnthropicTool;
    }
    // 3. Transform from OpenAI format (type: 'function', function: {...}) to Anthropic format
    if (tool.type === 'function' && tool.function) {
      return {
        name: tool.function.name,
        description: tool.function.description,
        input_schema: tool.function.parameters || {
          type: 'object',
          properties: {},
          required: [],
        },
      };
    }
    // 4. Fallback: assume it's missing input_schema, use parameters if available
    return {
      name: tool.name,
      description: tool.description,
      input_schema: tool.parameters ||
        tool.input_schema || {
          type: 'object',
          properties: {},
          required: [],
        },
    };
  });
}

/**
 * Handle Anthropic HTTP error responses by throwing a descriptive error.
 * Status-specific messages use keywords that route.ts error matching expects.
 * @param retryAfter Optional value of the 'retry-after' response header for 429 responses.
 */
function handleAnthropicHttpError(
  status: number,
  errorText: string,
  retryAfter?: string | null,
): never {
  if (status === 401) {
    throw new Error('Anthropic authentication error (401): Please check your API key.');
  } else if (status === 402) {
    throw new Error('Anthropic insufficient credits (402): Please upgrade your plan.');
  } else if (status === 403) {
    throw new Error(
      'Anthropic API permission denied (403): Your API key may not have access to this resource.',
    );
  } else if (status === 404) {
    throw new Error(`Anthropic not found (404): ${errorText}`);
  } else if (status === 413) {
    throw new Error('Anthropic API request too large (413): Maximum request size is 32 MB.');
  } else if (status === 429) {
    throw new Error(
      `Anthropic rate limit exceeded (429). ${retryAfter ? `Retry after ${retryAfter} seconds.` : 'Please try again later.'}`,
    );
  } else if (RETRYABLE_HTTP_STATUS_CODES.has(status)) {
    throw new Error(`Anthropic API service error (${status}): Please try again later.`);
  } else if (status === 529) {
    throw new Error('Anthropic API overloaded (529): Please try again later.');
  } else {
    throw new Error(`Anthropic API error: ${status} ${errorText}`);
  }
}

export class AnthropicProvider extends BaseLLMProvider {
  getDefaultBaseUrl(): string {
    return 'https://api.anthropic.com/v1';
  }

  protected override getHeaders(): Record<string, string> {
    return {
      'Content-Type': 'application/json',
      'x-api-key': this.apiKey,
      'anthropic-version': '2023-06-01', // Required API version (supports all Claude models)
      // Enable Claude 4.5/4.6 advanced features:
      // - output-128k-2025-02-19: Extended output limits for deep thinking
      // - context-1m-2025-08-07: 1M token context window
      // - interleaved-thinking-2025-05-14: Chain of thought with tool use
      'anthropic-beta':
        'output-128k-2025-02-19,context-1m-2025-08-07,interleaved-thinking-2025-05-14',
    };
  }

  async sendRequest(request: LLMProviderRequest): Promise<LLMProviderResponse> {
    const url = `${this.baseUrl}/messages`;

    // Convert messages format for Anthropic, preserving tool_use/tool_result blocks
    const messages = mapMessagesToAnthropic(
      request.messages.filter((msg) => msg.role !== 'system'),
      request.usePromptCache,
    );

    const systemMessage = request.messages.find((msg) => msg.role === 'system');

    // Build system content with cache_control if caching is enabled
    let systemContent: unknown = undefined;
    if (systemMessage) {
      if (request.usePromptCache) {
        systemContent = [
          {
            type: 'text',
            text: systemMessage.content,
            cache_control: { type: 'ephemeral' },
          },
        ];
      } else {
        systemContent = systemMessage.content;
      }
    }

    // Transform tools from OpenAI format to Anthropic format if needed
    const anthropicTools = request.tools
      ? transformTools(request.tools as OpenAITool[])
      : undefined;

    const body: Record<string, unknown> = {
      model: request.model,
      max_tokens: request.max_tokens || 16384, // Increased for Claude 4.5 quality outputs
      messages,
      ...(request.temperature !== undefined && { temperature: request.temperature }),
      ...(anthropicTools && { tools: anthropicTools }),
      ...(request.thinking && { thinking: request.thinking }),
      ...(request.effort && { effort: request.effort }),
    };

    if (systemContent) {
      body['system'] = systemContent;
    }

    try {
      const response = await fetch(url, {
        method: 'POST',
        headers: this.getHeaders(),
        body: JSON.stringify(body),
      });

      if (!response.ok) {
        let errorText: string;
        let errorData: unknown;
        try {
          errorText = await response.text();
          errorData = JSON.parse(errorText);
        } catch {
          errorText = response.statusText;
          errorData = { status: response.status };
        }

        logger.error(
          {
            status: response.status,
            error: errorText,
            errorData,
            model: request.model,
          },
          'Anthropic API error',
        );

        handleAnthropicHttpError(response.status, errorText, response.headers.get('retry-after'));
      }

      const data = await response.json();

      // Check stop_reason for error cases
      const stopReason = data.stop_reason;
      if (stopReason === 'max_tokens') {
        logger.warn(
          { model: request.model, stopReason },
          'Anthropic response was truncated due to max_tokens limit',
        );
      } else if (stopReason === 'refusal') {
        logger.warn({ model: request.model, stopReason }, 'Anthropic request was refused');
      }

      // Extract text content from content array
      // Concatenate ALL text blocks (web search responses may have multiple text blocks)
      const textBlocks = (data.content || []).filter(
        (block: { type: string }) => block.type === 'text',
      );
      const textContent = textBlocks.map((block: { text: string }) => block.text || '').join('');

      // Extract citations from text blocks (Anthropic web_search returns citations in text blocks)
      const citations: Array<{
        type: string;
        cited_text: string;
        title?: string;
        url?: string;
      }> = [];
      for (const block of textBlocks) {
        const typedBlock = block as {
          citations?: Array<{ type: string; cited_text: string; title?: string; url?: string }>;
        };
        if (typedBlock.citations && Array.isArray(typedBlock.citations)) {
          citations.push(...typedBlock.citations);
        }
      }

      // Extract web_search_tool_result blocks for passing search results to the client
      const searchResultBlocks = (data.content || []).filter(
        (block: { type: string }) => block.type === 'web_search_tool_result',
      );

      // Extract tool_use blocks (client-executed tools) and map to OpenAI-format tool_calls
      // Note: server_tool_use blocks (like web_search) are server-executed and do NOT
      // need client-side execution - Anthropic handles them internally
      const toolUseBlocks = (data.content || []).filter(
        (block: { type: string }) => block.type === 'tool_use',
      );
      const toolCalls =
        toolUseBlocks.length > 0
          ? toolUseBlocks.map(
              (block: { id: string; name: string; input: unknown }, index: number) => ({
                id: block.id,
                type: 'function' as const,
                index,
                function: {
                  name: block.name,
                  arguments: JSON.stringify(block.input || {}),
                },
              }),
            )
          : undefined;

      // Map Anthropic stop_reason to OpenAI finish_reason
      // server_tool_use stop_reason means Anthropic executed tools server-side and
      // returned results in the same response - this is a completed turn, map to 'stop'
      const finishReason =
        stopReason === 'tool_use' ? 'tool_calls' : stopReason === 'end_turn' ? 'stop' : stopReason;

      // Build response with optional web search metadata
      const result: LLMProviderResponse = {
        content: textContent,
        model: data.model || request.model,
        promptTokens: data.usage?.input_tokens || 0,
        completionTokens: data.usage?.output_tokens || 0,
        totalTokens: (data.usage?.input_tokens || 0) + (data.usage?.output_tokens || 0),
        finishReason,
        cacheCreationInputTokens: data.usage?.cache_creation_input_tokens,
        cachedInputTokens: data.usage?.cache_read_input_tokens,
        tool_calls: toolCalls,
      };

      // Attach web search citations and search results as extended metadata
      if (citations.length > 0) {
        result.citations = citations;
      }
      if (searchResultBlocks.length > 0) {
        result.search_results = searchResultBlocks;
      }

      return result;
    } catch (error) {
      logger.error({ error, model: request.model }, 'Anthropic request failed');
      throw error;
    }
  }
  async streamRequest(request: LLMProviderRequest): Promise<ReadableStream> {
    const url = `${this.baseUrl}/messages`;

    // Convert messages to Anthropic format, preserving tool_use/tool_result blocks
    // Pass usePromptCache so message-body cache_control is applied (matching sendRequest)
    const messages = mapMessagesToAnthropic(
      request.messages.filter((msg) => msg.role !== 'system'),
      request.usePromptCache,
    );

    const systemMessage = request.messages.find((msg) => msg.role === 'system');

    // Transform tools from OpenAI format to Anthropic format if needed
    const anthropicTools = request.tools
      ? transformTools(request.tools as OpenAITool[])
      : undefined;

    const body: Record<string, unknown> = {
      model: request.model,
      max_tokens: request.max_tokens || 16384, // Increased for Claude 4.5 quality outputs
      messages,
      stream: true,
      ...(request.temperature !== undefined && { temperature: request.temperature }),
      ...(anthropicTools && { tools: anthropicTools }),
      ...(request.thinking && { thinking: request.thinking }),
      ...(request.effort && { effort: request.effort }),
    };

    // Apply prompt cache on system message (matching sendRequest behavior)
    if (systemMessage) {
      if (request.usePromptCache) {
        body['system'] = [
          {
            type: 'text',
            text: systemMessage.content,
            cache_control: { type: 'ephemeral' },
          },
        ];
      } else {
        body['system'] = systemMessage.content;
      }
    }

    const response = await fetch(url, {
      method: 'POST',
      headers: this.getHeaders(),
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      let errorText: string;
      let errorData: unknown;
      try {
        errorText = await response.text();
        errorData = JSON.parse(errorText);
      } catch {
        errorText = response.statusText;
        errorData = { status: response.status };
      }

      logger.error(
        {
          status: response.status,
          error: errorText,
          errorData,
          model: request.model,
        },
        'Anthropic streaming API error',
      );

      handleAnthropicHttpError(response.status, errorText, response.headers.get('retry-after'));
    }

    if (!response.body) {
      throw new Error('No response body for streaming request');
    }

    return response.body;
  }
}

/**
 * Map messages to Anthropic format, converting tool_calls/tool_call_id
 * to Anthropic's tool_use/tool_result content blocks.
 */
type ContentPart = { type: string; text?: string; image_url?: { url: string; detail?: string } };

function buildAnthropicContentBlocks(
  text: string,
  multimodalContent?: ContentPart[],
): Array<Record<string, unknown>> {
  if (!multimodalContent || multimodalContent.length === 0) {
    return [{ type: 'text', text }];
  }

  return multimodalContent.map((part) => {
    if (part.type === 'image_url' && part.image_url?.url) {
      const url = part.image_url.url;
      // Parse data URLs: data:<mediaType>;base64,<data>
      const dataUrlMatch = url.match(/^data:([^;]+);base64,(.+)$/);
      if (dataUrlMatch) {
        return {
          type: 'image',
          source: {
            type: 'base64',
            media_type: dataUrlMatch[1] as string,
            data: dataUrlMatch[2] as string,
          },
        };
      }
      // Fall back to URL type for remote images
      return { type: 'image', source: { type: 'url', url } };
    }
    return { type: 'text', text: part.text ?? '' };
  });
}

function mapMessagesToAnthropic(
  messages: Array<{
    role: string;
    content: string;
    multimodal_content?: unknown[];
    tool_calls?: unknown[];
    tool_call_id?: string;
  }>,
  usePromptCache?: boolean,
): Array<Record<string, unknown>> {
  const result: Array<Record<string, unknown>> = [];

  for (let i = 0; i < messages.length; i++) {
    const msg = messages[i];
    if (!msg) continue;
    const isLast = i === messages.length - 1;

    // Tool role messages → Anthropic tool_result inside a user message
    if (msg.role === 'tool' && msg.tool_call_id) {
      // Anthropic requires tool_result blocks inside a role:"user" message.
      // Collect consecutive tool messages into a single user message.
      const toolResults: Array<Record<string, unknown>> = [];
      let j = i;
      while (j < messages.length) {
        const jMsg = messages[j];
        if (!jMsg || jMsg.role !== 'tool' || !jMsg.tool_call_id) break;
        toolResults.push({
          type: 'tool_result',
          tool_use_id: jMsg.tool_call_id,
          content: jMsg.content || '',
        });
        j++;
      }
      if (toolResults.length > 0) {
        // Re-evaluate whether this batch is the last in the conversation.
        // isLast was computed before the inner while-loop, so it can be wrong
        // when a conversation ends with tool-result messages.
        const batchIsLast = j === messages.length;
        if (batchIsLast && usePromptCache && toolResults.length > 0) {
          // Apply cache_control to the last tool_result block in the batch
          const lastResult = toolResults[toolResults.length - 1]!;
          (lastResult as Record<string, unknown>)['cache_control'] = { type: 'ephemeral' };
        }
        result.push({
          role: 'user',
          content: toolResults,
        });
      }
      // Skip the messages we just consumed (loop will increment i)
      i = j - 1;
      continue;
    }

    // Assistant messages with tool_calls → Anthropic tool_use content blocks
    if (msg.role === 'assistant' && Array.isArray(msg.tool_calls) && msg.tool_calls.length > 0) {
      const contentBlocks: Array<Record<string, unknown>> = [];
      // Add text content if present
      if (msg.content && msg.content.trim().length > 0) {
        contentBlocks.push({ type: 'text', text: msg.content });
      }
      // Convert each tool_call to a tool_use block
      for (const tc of msg.tool_calls as Array<{
        id?: string;
        function?: { name?: string; arguments?: string };
      }>) {
        contentBlocks.push({
          type: 'tool_use',
          id: tc.id || `call_${Date.now()}`,
          name: tc.function?.name || 'unknown',
          input: tc.function?.arguments
            ? (() => {
                try {
                  return JSON.parse(tc.function!.arguments as string);
                } catch {
                  return {};
                }
              })()
            : {},
        });
      }
      result.push({ role: 'assistant', content: contentBlocks });
      continue;
    }

    // Skip empty content messages (unless there are image attachments)
    const multimodalParts = msg.multimodal_content as ContentPart[] | undefined;
    const hasImages = multimodalParts?.some((p) => p.type === 'image_url');
    if (!hasImages && (!msg.content || msg.content.trim().length === 0)) {
      continue;
    }

    // Normal messages (possibly multimodal)
    let contentValue: unknown;
    if (hasImages) {
      // Build Anthropic content blocks from multimodal parts
      const blocks = buildAnthropicContentBlocks(msg.content, multimodalParts);
      contentValue =
        usePromptCache && isLast
          ? blocks.map((b, idx) =>
              idx === blocks.length - 1 ? { ...b, cache_control: { type: 'ephemeral' } } : b,
            )
          : blocks;
    } else if (usePromptCache && isLast) {
      contentValue = [{ type: 'text', text: msg.content, cache_control: { type: 'ephemeral' } }];
    } else {
      contentValue = msg.content;
    }

    result.push({
      role: msg.role === 'tool' ? 'user' : msg.role,
      content: contentValue,
    });
  }

  return result;
}
